BusyList/ITaskRepository.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System.Collections.Generic;
+
+namespace BusyList
+{
+    public record AddTaskData(string Description);
+
+    public record TaskItem(int Id, string Description);
+
+    public interface ITaskRepository
+    {
+        /// <summary>
+        /// Get a task from the repository that matches the id parameter. 
+        /// </summary>
+        /// <param name="id"></param>
+        /// <returns>The first task that matches the id parameter.</returns>
+        TaskItem GetTaskById(int id);
+
+        /// <summary>
+        /// Add a new task to the repository.
+        /// </summary>
+        /// <param name="data">Input data for the new task.</param>
+        /// <returns>The newly created task in the repository.</returns>
+        TaskItem AddTask(AddTaskData data);
+
+        /// <summary>
+        /// Marks the task as deleted in the repository.
+        /// </summary>
+        /// <param name="task"></param>
+        /// <returns>The updated task.</returns>
+        TaskItem DeleteTask(TaskItem task);
+
+        /// <summary>
+        /// Gets all the tasks in the repository.
+        /// </summary>
+        /// <returns>All tasks in the repository.</returns>
+        IEnumerable<TaskItem> GetAll();
+
+        /// <summary>
+        /// Update an existing task in the repository.
+        /// </summary>
+        /// <param name="task"></param>
+        /// <returns>The updated task.</returns>
+        TaskItem UpdateTask(TaskItem task);
```

Comment:
How can we update a task if we cannot change its properties since its immutable?

Reply:
They can't be updated in your classical sense, but using the `with {}` syntax a copy can be created that has new values. This will work outside of EF (as we can simply replace the old object with a new, since there is no change tracking).

If we move to EF later, we will have to make TaskItem mutable.